src/hooks/useProcessingIds.ts

```diff
@@ -0,0 +1,59 @@
+import { useState, useMemo } from 'react';
+
+export type UseProcessingReturn = {
+  processingIds: Set<number>;
+  addProcessing: (id: number) => void;
+  removeProcessing: (id: number) => void;
+  addManyProcessing: (ids: number[]) => void;
+  isProcessing: (id: number) => boolean;
+};
+
+export const useProcessingIds = (): UseProcessingReturn => {
+  const [processingIds, setProcessingIds] = useState<Set<number>>(new Set());
+
+  const startProcessing = (id: number) =>
+    setProcessingIds(prev => {
+      const newSet = new Set(prev);
+
+      newSet.add(id);
+
+      return newSet;
+    });
+
+  const removeProcessing = (id: number) =>
+    setProcessingIds(prev => {
+      const newSet = new Set(prev);
+
+      newSet.delete(id);
+
+      return newSet;
+    });
+
+  const addManyProcessing = (ids: number[]) =>
+    setProcessingIds(prev => {
+      if (ids.length === 0) {
+        return prev;
+      }
+
+      const newSet = new Set(prev);
+
+      ids.forEach(id => newSet.add(id));
+
+      return newSet;
+    });
```

Comment:
You can unite two functions into one

```suggestion
  const addProcessing = (ids: number | number[]) =>
    const idsToProcess = Array.isArray(ids) ? ids : [ids];
    
    setProcessingIds(prev => {
      if (idsToProcess.length === 0) {
        return prev;
      }

      const newSet = new Set(prev);

      idsToProcess.forEach(id => newSet.add(id));

      return newSet;
    });
```